text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,1062 @@
+take several inputs (the parameters to the trait) which may or may not
+be related to the actual type definition in question.)
+
+For the time being, the current implementation takes a pragmatic
+approach based on heuristics. It tries to avoid adding edges to the
+region graph in various common scenarios, and in the end falls back to
+enforcing conditions that may be stricter than necessary, but which
+certainly suffice. We have not yet encountered an example in practice
+where the current implementation rules do not suffice.
+
+### The WF relation
+
+This section describes the "well-formed" relation. In
+[previous RFCs][RFC 192], this was combined with the outlives
+relation. We separate it here for reasons that shall become clear when
+we discuss WF conditions on impls.
+
+The WF relation is really pretty simple: it just says that a type is
+"self-consistent". Typically, this would include validating scoping
+(i.e., that you don't refer to a type parameter `X` if you didn't
+declare one), but we'll take those basic conditions for granted.
+
+    WfScalar:
+      --------------------------------------------------
+      R ⊢ scalar WF
+
+    WfParameter:
+      --------------------------------------------------
+      R ⊢ X WF                  // where X is a type parameter
+
+    WfTuple:
+      ∀i. R ⊢ Ti WF
+      --------------------------------------------------
+      R ⊢ (T0..Tn) WF
+
+    WfNominalType:
+      ∀i. R ⊢ Pi Wf             // parameters must be WF,
+      C = WhereClauses(Id)      // and the conditions declared on Id must hold...
+      R ⊢ [P0..Pn] C            // ...after substituting parameters, of course
+      --------------------------------------------------
+      R ⊢ Id<P0..Pn> WF
+
+    WfReference:
+      R ⊢ T WF                  // T must be WF
+      R ⊢ T: 'x                 // T must outlive 'x
+      --------------------------------------------------
+      R ⊢ &'x T WF
+
+    WfSlice:
+      R ⊢ T WF
+      R ⊢ T: Sized
+      --------------------------------------------------
+      [T] WF
+
+    WfProjection:
+      ∀i. R ⊢ Pi WF             // all components well-formed
+      R ⊢ <P0: Trait<P1..Pn>>   // the projection itself is valid
+      --------------------------------------------------
+      R ⊢ <P0 as Trait<P1..Pn>>::Id WF
+
+#### WF checking and higher-ranked types
+
+There are two places in Rust where types can introduce lifetime names
+into scope: fns and trait objects. These have somewhat different rules
+than the rest, simply because they modify the set `R` of bound
+lifetime names. Let's start with the rule for fn types:
+
+    WfFn:
+      ∀i. R, r.. ⊢ Ti WF
+      --------------------------------------------------
+      R ⊢ for<r..> fn(T1..Tn) -> T0 WF
+
+Basically, this rule adds the bound lifetimes to the set `R` and then
+checks whether the argument and return type are well-formed. We'll see
+in the next section that means that any requirements on those types
```

Comment:
The next section is "WF checking a trait reference", so I don't know what exactly you are referring to here.